botocore/credentials.py

```diff
@@ -791,6 +829,212 @@ def _assume_role_base_kwargs(self, config):
+            raise SAMLError(
+                detail='Failed to login at %s' % config['saml_endpoint'])
+        idp_roles = self._parse_roles(assertion)
+        role = self.role_selector(config.get('role_arn'), idp_roles)
+        if not role:
+            raise SAMLError(detail='Unable to choose role "%s" from %s' % (
+                config.get('role_arn'), [r['RoleArn'] for r in idp_roles]))
+        role['SAMLAssertion'] = assertion
+        return role
+
+    def _parse_roles(self, assertion):
+        attribute = '{urn:oasis:names:tc:SAML:2.0:assertion}Attribute'
+        attr_value = '{urn:oasis:names:tc:SAML:2.0:assertion}AttributeValue'
+        awsroles = []
+        root = ET.fromstring(base64.b64decode(assertion))
+        for attr in root.getiterator(attribute):
+            if attr.get('Name') == \
+                    'https://aws.amazon.com/SAML/Attributes/Role':
+                for value in attr.getiterator(attr_value):
+                    parts = value.text.split(',')
+                    # Deals with "role_arn,pricipal_arn" or its reversed order
+                    if 'saml-provider' in parts[0]:
+                        role = {'PrincipalArn': parts[0], 'RoleArn': parts[1]}
+                    else:
+                        role = {'PrincipalArn': parts[1], 'RoleArn': parts[0]}
+                    awsroles.append(role)
+        return awsroles
+
+
+class SAMLAuthenticator(object):
+    def is_suitable(self, config):
+        """Return True if this instance intends to perform authentication.
+
+        :type config: dict
+        :param config: It is the profile dictionary loaded from user's profile,
+            i.e. {'saml_endpoint': 'https://...', 'saml_provider': '...', ...}
+        """
+        raise NotImplemented()
+
+    def authenticate(self, config, username_prompter, password_prompter):
+        """Returns SAML assertion when login succeeds, or None otherwise"""
+        raise NotImplemented()
+
+
+class SAMLGenericFormsBasedAuthenticator(SAMLAuthenticator):
+    username_field = 'username'
+    password_field = 'password'
+
+    def is_suitable(self, config):
+        return config.get('saml_authentication_type') == 'form'
+
+    def authenticate(self, config, username_prompter, password_prompter):
+        verify = True
+        endpoint = config['saml_endpoint']
+        login_form = self._get_form(requests.get(endpoint, verify=verify).text)
+        if login_form is None:
+            raise SAMLError(detail='Login form is not found in %s' % endpoint)
+        form_action = urljoin(endpoint, login_form.attrib.get('action', ''))
+        remind = ''
+        if not form_action.lower().startswith('https://'):
+            remind = ("Your IdP is not using secure HTTPS connection. "
+                      "Abort if you don't want to proceed.\n")
+            logger.warn(remind)
+        if not config.get('saml_username'):
+            config['saml_username'] = username_prompter(remind + "Username: ")
```

Comment:
Well, actually I knew it was a side effect when I wrote it like this at the first place. I thought it was harmless (or even arguably desirable), because we already asked user to input it once, so we could save it here just in case we would need it later. That being said, I don't mind to remove it in my next commit.